src/main.rs

```diff
@@ -135,15 +140,18 @@ where
     }
 
     // Setup channels that we'll use to signal to spawned producers that an exit has occurred elsewhere so they should
-    // exit.
-    let (exit_sender, exit_receiver) = channel(1);
+    // exit, and a tokio task set to track all our spawned tasks.
```

Comment:
This `run_onion_messenger` function is getting a bit long (especially since my coming PR will be adding to it too!), could we split it into a few functions? Perhaps it could be refactored into something like a OnionMessengerRelayer struct with one main `run` method that calls all the other methods. Thoughts? Maybe could leave this for a later PR though

Reply:
I think that we can be cheeky and leave as is for now, since your PR is the last piece of the puzzle for forwaring OMs. 

Really like the idea of passing in a set of abstract tasks with a `run` fucntion that handles all the setup.